### ISSUE TITLE
Kotlin DSL resolution strategy is ambiguous

### ISSUE THREAD

**Issue by jeffdgr8** [ben-manes/gradle-versions-plugin#808]
The Gradle Kotlin DSL example in the readme:
```kotlin
tasks.withType<DependencyUpdatesTask> {
  resolutionStrategy {
    componentSelection {
      all {
        if (isNonStable(candidate.version) && !isNonStable(currentVersion)) {
          reject("Release candidate")
        }
      }
    }
  }
}
```
doesn't work without an explicit argument name in the `all()` function call.

As the code is written, Android Studio resolves `all { ... }` to the more general [`all(ruleSource: Any)`](https://github.com/ben-manes/gradle-versions-plugin/blob/610f5b2e81923cbbd4e4df0ae6a91bfc414d794e/gradle-versions-plugin/src/main/kotlin/com/github/benmanes/gradle/versions/updates/resolutionstrategy/ComponentSelectionRulesWithCurrent.kt#L37), rather than the intended [`all(selectionAction: Action<in ComponentSelectionWithCurrent>)`](https://github.com/ben-manes/gradle-versions-plugin/blob/610f5b2e81923cbbd4e4df0ae6a91bfc414d794e/gradle-versions-plugin/src/main/kotlin/com/github/benmanes/gradle/versions/updates/resolutionstrategy/ComponentSelectionRulesWithCurrent.kt#L16-L18).

Adding the named argument fixes the code:
```kotlin
tasks.withType<DependencyUpdatesTask> {
  resolutionStrategy {
    componentSelection {
      all(selectionAction = {
        if (isNonStable(candidate.version) && !isNonStable(currentVersion)) {
          reject("Release candidate")
        }
      })
    }
  }
}
```

This seems to be an IDE issue at its core, as the code runs successfully without the named arguments. But both Android Studio and IntelliJ IDEA both can't resolve the `candidate` and `reject()` code and traverse to the incorrect `all()` function when cmd+clicking.

**Comment by ben-manes** [comment]
yeah, it sounds like an IDE issue since I am [using that](https://github.com/ben-manes/caffeine/blob/fef5e2d06e4e4f6f7a400571456ab092dc0776a2/gradle/plugins/src/main/kotlin/lifecycle/dependency-versions-caffeine-conventions.gradle.kts#L7-L22) fine at command-line. I guess open the issue with IDEA?

**Comment by jeffdgr8** [comment]
Thanks for the quick response @ben-manes. What version of Kotlin are you using in your code? Perhaps this is a regression in Kotlin 1.9?

**Comment by jeffdgr8** [comment]
Looks like you're using Kotlin 1.9.10. 🤔 Do you not see errors for this code in your IDE?

**Comment by ben-manes** [comment]
Gradle runs on 1.8.x until the next major release, iirc. I only know that because I get qodana warnings on 1.9 deprecations in the build files, but its not yet what Gradle runs as (and had found https://github.com/gradle/gradle/issues/25517#issue-1776410738 to understand why).

**Comment by ben-manes** [comment]
> Looks like you're using Kotlin 1.9.10. 🤔 Do you not see errors for this code in your IDE?

I use Eclipse for development and VSCode for build files. Unfortunately IntelliJ goes into a death spiral when trying to edit kotlin build scripts so that it is faster to iterate in a text editor and cli than using their tooling. I kept trying when migrating to a kotlin build but it was unusable, not sure if that is the common experience or due to heavy build refactoring?

**Comment by jeffdgr8** [comment]
I created a YouTrack issue: https://youtrack.jetbrains.com/issue/KTIJ-26964/Function-call-in-Kotlin-Gradle-script-resolves-to-incorrect-overload

**Comment by ben-manes** [comment]
Closing as acknowledged by Jetbrains as a known regression (fixed for Gradle 8.4 iiuc).
